### PR TITLE
[layouts] add icons for items and use those in the items panel

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This PR adds icons for all layout items, and display those in the layout's items panel. The panel now looks a lot more like the main window's layers panel, which is nice. It also helps spotting the item(s) you are looking for in complex layouts.

Screenshot:
screenshot from 2017-12-20 16-33-29

I've implemented this by adding an icon() function in the QgsLayoutItem class, which can be overwritten by individual items. It'll allow for plugins to have icons of their own. It also opens the door to do stuff like using the rendering cache of a map item as its icon instead of the generic icon, etc.

@nyalldawson , review most appreciated, as always.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
Grayscaled the icons:
screenshot from 2017-12-20 17-26-42

**Comment by nirvn** [comment]
@nyalldawson , icons not showing up for HTML and attribute table items. Not sure why.

**Comment by nirvn** [comment]
Are are the two missing icons in action:
screenshot from 2017-12-20 17-47-39

**Comment by nyalldawson** [comment]
Looks beautiful!

Looking at that last screenshot... should we grayscale the lock/eye icon too? (We already have a greyscale lock)

**Comment by nirvn** [comment]
@nyalldawson , not a bad idea.

**Comment by nirvn** [comment]
@nyalldawson , on the item name front, I think "Scale Bar" should have bar lowercased. Also, by default "<legend>" should have a capital L.

**Comment by nyalldawson** [comment]
>  Also, by default "" should have a capital L.

What do you mean here?

**Comment by nyalldawson** [comment]
Oh, legend?

**Comment by nirvn** [comment]
@nyalldawson , yep, somehow my smartphone are "legend" :)

**Comment by borysiasty** [comment]
For html-speaking people: it was &amp;lt;legend&amp;gt; ;)

File: src/core/layout/qgslayoutitem.h [hunk 1]
```diff
@@ -292,12 +293,17 @@ class CORE_EXPORT QgsLayoutItem : public QgsLayoutObject, public QGraphicsRectIt
...
+     * Returns a unique graphics item type identifier.
      *
      * Plugin based subclasses should return an identifier greater than QgsLayoutItemRegistry::PluginItem.
      */
     int type() const override;
 
+    /**
+     * Returns the item's icon.
+     */
+    virtual QIcon icon() { return QIcon(); }
```
Comment (nyalldawson) [inline src/core/layout/qgslayoutitem.h]: const

Comment (nyalldawson) [reply src/core/layout/qgslayoutitem.h]: Should we provide a default here? (e.g. rectangle shape icon). It's always rather ugly when a list has missing icons, and when Python code starts adding extra item types I can see them skipping the icon...

**Comment by nyalldawson** [comment]
Another nice things about this is that these icons should show in the layout item combo boxes... which will help give a visual clue to those widgets too!

**Comment by nirvn** [comment]
@nyalldawson , comments addressed.

**Comment by nyalldawson** [comment]
Ok, merge when you're ready!

**Pull request merged**